### PR TITLE
Add baseball example in numpyro

### PR DESCRIPTION
**Pull request by neerajprad**
This adds the baseball example to numpyro, following pretty much the same style as in Pyro. I am seeing a speedup of ~20-40X speedup in some of these models.

Note that inference fails for the `no-pooling` model unless we disable fastmath mode, and I spent quite a long time trying to pinpoint where the bug lay. 

Addresses: #90. Note that I have not benchmarked against Stan, and will be doing that separately.

### CODE REVIEW

File: numpyro/examples/baseball.py [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+        rng, rng_predict = random.split(random.PRNGKey(i))
+        zs = run_inference(model, at_bats, hits, random.PRNGKey(i), args)
+        predict(model, at_bats, hits, zs, rng_predict, player_names)
+        predict(model, season_at_bats, season_hits, zs, rng_predict, player_names, train=False)
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="Baseball batting average using HMC")
+    parser.add_argument("-n", "--num-samples", nargs="?", default=3000, type=int)
+    parser.add_argument("--num-warmup-steps", nargs='?', default=1500, type=int)
```
Comment (fehiepsi) [inline numpyro/examples/baseball.py]: Not for this PR: should we change this arg in hmc to `warmup_steps`?

File: numpyro/examples/baseball.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+import numpyro.distributions as dist
+from numpyro.examples.datasets import BASEBALL, load_dataset
+from numpyro.handlers import sample, seed, substitute, trace
+from numpyro.hmc_util import initialize_model
+from numpyro.mcmc import hmc
+from numpyro.util import fori_collect
+
+"""
+Original example from Pyro: 
+https://github.com/pyro-ppl/pyro/blob/dev/examples/baseball.py
```
Comment (fehiepsi) [inline numpyro/examples/baseball.py]: lint: trailing spaces at this line and the above line

File: numpyro/examples/baseball.py (hunk 1 above)
Comment (neerajprad) [reply numpyro/examples/baseball.py]: Sure, I think there was no other reason to name it `num_warmup_steps` than to have a similar name to `num_samples`. Maybe we can also rename it as `num_warmup`. :)

File: numpyro/examples/datasets.py [hunk 3]
```diff
@@ -86,8 +114,9 @@ def load_dataset(dset, batch_size=None, split='train', shuffle=True):
     def init():
         return num_records // batch_size, np.random.permutation(idxs) if shuffle else idxs
 
-    def get_batch(i, idxs):
+    def get_batch(i=0, idxs=idxs):
         ret_idx = lax.dynamic_slice_in_dim(idxs, (i + 1) * batch_size, batch_size)
-        return tuple(lax.index_take(a, (ret_idx,), axes=(0,)) for a in arrays)
+        return tuple(lax.index_take(a, (ret_idx,), axes=(0,)) if isinstance(a, DeviceArray)
```
[the remark this answers is not in the document]
Comment (neerajprad) [reply numpyro/examples/datasets.py]: The issue here is that `get_batch` might be called from inside a while loop and therefore needs to be traceable which native python indexing ops won't be. I honestly don't remember why I had `i+1` in there, changed to `i`. Thanks for catching that.

File: numpyro/examples/baseball.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+                  hits)
+    if not train:
+        model = substitute(model, z)
+        model_trace = trace(model).get_trace(at_bats, hits)
+        log_joint = 0.
+        for site in model_trace.values():
+            site_log_prob = site['fn'].log_prob(site['value'])
+            log_joint = log_joint + onp.apply_over_axes(np.sum,
+                                                        site_log_prob,
+                                                        range(1, np.ndim(site_log_prob)))
```
[the remark this answers is not in the document]
Comment (neerajprad) [reply numpyro/examples/baseball.py]: Done.

**Comment by neerajprad** [comment]
>>> It is really nice (and surprised to me at the first look) to see prediction for the whole zs at the same time! Do you have an idea how to support it out of box? I think that it is doable when we support plate notation but the problem can be more complicated than I thought. I would be nice if you can add some todos or notes at places which we can improve in the future:

I just added some TODOs. You are right that this can be done easily once we have `pyro.plate` - we just need to wrap the model inside an outermost `plate` when replaying. The current logic is kind of hacky and while it works because these models are simple, I don't think it's something that can be easily replicated for most complex models. I would have liked to test this in Pyro, but we have more complications there because I am not sure how this will play with data subsampling in SVI.

File: numpyro/examples/datasets.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (fehiepsi) [reply numpyro/examples/datasets.py]: > The issue here is that get_batch might be called from inside a while loop

Got it!

**Comment by fehiepsi** [comment]
Yeah, I think that it is worth to think about it. It is pretty slow to loop over all traces to get predictions in Pyro, even for small models (which I made to replicate Statistical Rethinking book).

> I am not sure how this will play with data subsampling in SVI.

I'll think about it too when it is needed. :)

**Comment by fehiepsi** [comment]
@neerajprad I get the same issue with the old version of jaxlib. Please update your jaxlib to the newer version. :)

**Pull request merged**